ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -567,102 +627,104 @@ impl<'a> Assemble<'a> {
                     &mut self.asm,
                     iidx,
                     [RegConstraint::InputOutput(lhs), RegConstraint::Input(rhs)],
                 );
-                match size {
-                    1 => dynasm!(self.asm; or Rb(lhs_reg.code()), Rb(rhs_reg.code())),
-                    2 => dynasm!(self.asm; or Rw(lhs_reg.code()), Rw(rhs_reg.code())),
-                    4 => dynasm!(self.asm; or Rd(lhs_reg.code()), Rd(rhs_reg.code())),
-                    8 => dynasm!(self.asm; or Rq(lhs_reg.code()), Rq(rhs_reg.code())),
+                match byte_size {
+                    1..=8 => {
+                        // OK to ignore any undefined high-order bits here.
+                        dynasm!(self.asm; or Rq(lhs_reg.code()), Rq(rhs_reg.code()));
+                    }
                     _ => todo!(),
                 }
             }
             BinOp::SDiv => {
-                let size = lhs.byte_size(self.m);
+                let Ty::Integer(bit_size) = self.m.type_(lhs.tyidx(self.m)) else {
+                    unreachable!()
+                };
                 self.ra
                     .clobber_gp_regs_hack(&mut self.asm, iidx, &[Rq::RDX]);
-                let [_lhs_reg, rhs_reg] = self.ra.assign_gp_regs_avoiding(
+                let [lhs_reg, rhs_reg] = self.ra.assign_gp_regs_avoiding(
                     &mut self.asm,
                     iidx,
                     [
-                        // The quotient is stored in RAX. We don't care about the remainder stored
-                        // in RDX.
+                        // 64-bit signed division with idiv operates on RDX:RAX and stores the
+                        // quotient in RAX (we ignore the remainder put into RDX).
                         RegConstraint::InputOutputIntoReg(lhs, Rq::RAX),
                         RegConstraint::Input(rhs),
                     ],
                     RegSet::from(Rq::RDX),
                 );
-                // The dividend is hard-coded into DX:AX/EDX:EAX/RDX:RAX. However unless we have 128bit
-                // values or want to optimise register usage, we won't be needing this, and just zero out
-                // RDX.
-
-                // Signed division (idiv) operates on the DX:AX, EDX:EAX, RDX:RAX registers, so we
-                // use `cdq`/`cqo` to double the size via sign extension and store the result in
-                // DX:AX, EDX:EAX, RDX:RAX.
-                match size {
-                    // There's no `cwd` equivalent for byte-sized values, so we use `movsx`
-                    // (sign-extend) instead.
-                    1 => dynasm!(self.asm; movsx ax, al; idiv Rb(rhs_reg.code())),
-                    2 => dynasm!(self.asm; cwd; idiv Rw(rhs_reg.code())),
-                    4 => dynasm!(self.asm; cdq; idiv Rd(rhs_reg.code())),
-                    8 => dynasm!(self.asm; cqo; idiv Rq(rhs_reg.code())),
+                match bit_size {
```

Comment:
A lot of these `match`es are also `if/else` in disguise (but with more indentation).

Reply:
I used match to take advantage of the `1..=64` notation, but I will convert them to ifs.